File: python/dp/best_time_buy_sell_stock.py

# 买卖股票的最佳时机I
from typing import List


class Solution:
    # 暴力法，两层for循环遍历
    def maxProfit_1(self, prices: List[int]) -> int:
        profit = 0
        for i in range(len(prices) - 1):
            for j in range(i + 1, len(prices)):
                if prices[j] - prices[i] > profit:
                    profit = prices[j] - prices[i]
        return profit

    def maxProfit_2(self, prices: List[int]) -> int:
        if len(prices) == 0:
            return 0
        profit, minim = 0, prices[0]
        for i in range(1, len(prices)):
            if prices[i] < minim:
                minim = prices[i]
            profit = max(profit, prices[i] - minim)
        return profit

File: java/datastructure/array/leetcode/BestTimeBuySellStock.java
package datastructure.array.leetcode;

/**
 * @author roseduan
 * 买卖股票的最佳时机I
 */
public class BestTimeBuySellStock {
    
    /**
     * 暴力法，遍历寻找最大值
     */
    public int maxProfit1(int[] prices) {
        int max = 0;
        for (int i = 0; i < prices.length - 1; i++) {
            for (int j = i + 1; j < prices.length; j++) {
                max = Math.max(max, prices[j] - prices[i]);
            }
        }
        return max;
    }
    
    /**
     * 一次遍历
     */
    public int maxProfit2(int[] prices) {
        int res = 0, min = Integer.MAX_VALUE;
        for (int n : prices) {
            if (n < min) {
                min = n;
            }
            res = Math.max(res, n - min);
        }
        return res;
    }
}
